File: Leetcode/MinimumChangesToMakeAlternatingBinaryString.rs
// https://leetcode.com/problems/minimum-changes-to-make-alternating-binary-string

struct Solution {}

impl Solution {
    pub fn min_operations(s: String) -> i32 {
        (0..s.len())
            .fold(vec![vec![]; 2], |mut patterns, ind| {
                patterns[0].push((b'0' + (ind & 1) as u8) as char);
                patterns[1].push((b'0' + ((ind & 1) as u8) ^ 1) as char);
                patterns
            })
            .into_iter()
            .map(|pattern| {
                pattern
                    .into_iter()
                    .zip(s.chars())
                    .filter(|(c1, c2)| c1 != c2)
                    .count()
            })
            .min()
            .unwrap() as i32
    }
}

File: Leetcode/MinimumChangesToMakeAlternatingBinaryString.cpp
// https://leetcode.com/problems/minimum-changes-to-make-alternating-binary-string

class Solution {
   public:
    int minOperations(string s) {
        int case1 = 0, case2 = 0;
        for (size_t i = 0; i < s.length(); ++i) {
            case1 += (1 - (i & 1)) & (s[i] == '0');
            case1 += (i & 1) & (s[i] == '1');

            case2 += (1 - (i & 1)) & (s[i] == '1');
            case2 += (i & 1) & (s[i] == '0');
        }

        return min(case1, case2);
    }
};